### PR TITLE
[PM-863] Fix Organization Folders in EF Databases

### LINKED ISSUE #2561: Cannot see items in vault
### Steps To Reproduce

1. Log in to Bitwarden self-host environment using desktop client or Chrome extension
2. Click on Vault tab

### Expected Result

See all of my saved logins and secure notes

### Actual Result

There are no items to list.

### Screenshots or Videos

_No response_

### Additional Context

After installing self-hosted Bitwarden, I imported my passwords from Lastpass. A lot of them were duplicated for some reason. I was going through and editing/de-duplicating. After changing the name of one, I tried to move it to my Organization. It gave me an error about the cipher being out of date, and to re-sync. I tried to re-sync and it says Sync failed. I logged out and back in, and now my vault shows as being empty. As far as I can tell, it all still exists in the database. I have over 1800 entries in the Cipher table, my vault shows empty in the web UI, the desktop application, and the Chrome extension. I added my email to the the Admin users so that I could get to the admin console. It shows my users and Organization, but if I click on either user, it gives me a 404 and this error shows up in api.log:

```2023-01-10 16:28:27.463 +00:00 [ERR] The JSON value could not be converted to System.Collections.Generic.Dictionary`2[System.Guid,System.Guid]. Path: $ | LineNumber: 0 | BytePositionInLine: 4.
System.Text.Json.JsonException: The JSON value could not be converted to System.Collections.Generic.Dictionary`2[System.Guid,System.Guid]. Path: $ | LineNumber: 0 | BytePositionInLine: 4.
 ---> System.FormatException: The JSON value is not in a supported Guid format.
   at System.Text.Json.Utf8JsonReader.GetGuidNoValidation()
   at System.Text.Json.Serialization.Converters.GuidConverter.ReadAsPropertyNameCore(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonDictionaryConverter`3.<OnTryRead>g__ReadDictionaryKey|10_0(Utf8JsonReader& reader, ReadStack& state, <>c__DisplayClass10_0& )
   at System.Text.Json.Serialization.JsonDictionaryConverter`3.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TDictionary& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, Utf8JsonReader& reader, Exception ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at Bit.Infrastructure.EntityFramework.Repositories.Queries.UserCipherDetailsQuery.GetFolderId(Nullable`1 userId, Cipher cipher) in /source/src/Infrastructure.EntityFramework/Repositories/Queries/UserCipherDetailsQuery.cs:line 83
   at lambda_method420(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Bit.Infrastructure.EntityFramework.Repositories.CipherRepository.GetManyByUserIdAsync(Guid userId, Boolean withOrganizations) in /source/src/Infrastructure.EntityFramework/Repositories/CipherRepository.cs:line 353
   at Bit.Api.Controllers.SyncController.Get(Boolean excludeDomains) in /source/src/Api/Controllers/SyncController.cs:line 72
   at lambda_method87(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|26_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)```

### Githash Version

3997defe-dirty

### Environment Details

- Environment: Docker
- Hardware: Synology NAS

### Database Image

postgres:15

### Issue-Link

https://github.com/bitwarden/server/issues/2480

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

### PR DESCRIPTION
**Pull request by justindbaur**
## Type of change

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Resolves #2561

When moving a cipher from a personal vault to an organization one we use the `UserId` property for the user doing the move and the id used for the data on if it is favorited or in a folder. But since this cipher is being moved from personal to organization it's `UserId` column in the database will eventually be null. Before this change we were doing this early in the method. 

https://github.com/bitwarden/server/blob/61a0efbdfcacc54ab039a5695b78464135820f34/src/Infrastructure.EntityFramework/Vault/Repositories/CipherRepository.cs#L421-L423

This would rightfully change `cipher.UserId` to be null, but since `cipher.UserId` is used later in the method to be the key in the `Folder` and `Favorite` dictionary it would create json like: `{ "": "<valid guid>" }`. So with this fix I do that bit of code a bit later in the method after the `UserId` has been utilized. 

## Code changes

* **src/Infrastructure.EntityFramework/Repositories/Queries/UserCipherDetailsQuery.cs:** Make the `GetFolderId` logic more resilient to a possible JSON format issue that this bug caused. 
* **src/Infrastructure.EntityFramework/Vault/Repositories/CipherRepository.cs:** Construct folder and favorite JSON bodies in a more type safe manner. Most importantly don't set the `UserId` column to it's value until after the JSON bodies have made use of the UserId for it's keys. 

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by MGibson1 — COMMENTED** [review]
Can you please add tests of the issue to confirm resolution? I'm not able to follow the bug described very clearly

**Comment by justindbaur** [comment]
@MGibson1 Updated the PR body to have a better description and added a test. Let me know if you still don't follow.

File: src/Infrastructure.EntityFramework/Vault/Repositories/CipherRepository.cs [hunk 1]
```diff
@@ -447,32 +449,45 @@ public async Task ReplaceAsync(CipherDetails cipher)
...
                 }
                 else
                 {
                     if (cipher.Folders != null && cipher.Folders.Contains(cipher.UserId.Value.ToString()))
                     {
-                        var folders = CoreHelpers.LoadClassFromJsonData<Dictionary<Guid, bool>>(cipher.Favorites);
+                        var folders = CoreHelpers.LoadClassFromJsonData<Dictionary<Guid, Guid>>(cipher.Folders);
                         folders.Remove(cipher.UserId.Value);
-                        cipher.Favorites = JsonConvert.SerializeObject(folders);
+                        cipher.Favorites = JsonSerializer.Serialize(folders);
```
Comment (MGibson1) [inline src/Infrastructure.EntityFramework/Vault/Repositories/CipherRepository.cs:478]: This section is seriously confusing. Are we working with folders or favorites?

I have a feeling 478 should be `cipher.Folders`

**Review by MGibson1 — CHANGES_REQUESTED** [review]

File: src/Infrastructure.EntityFramework/Vault/Repositories/CipherRepository.cs (hunk 1 above)
Comment (justindbaur) [reply src/Infrastructure.EntityFramework/Vault/Repositories/CipherRepository.cs:478]: It definitely should be, I partially fixed it by switching it to `.Folders` but forgot that last little bit. Technically isn't a part of the original issue but something I found along the way.

**Review by MGibson1 — APPROVED** [review]

**Comment by MGibson1** [comment]
Adding needs QA here since I think we actually fixed a favorites bug, but if you think the integration test is sufficient go ahead and remove to merge

**Pull request merged**